File: server/relay.py
#!/usr/bin/env python3
"""
The relay that pairs a phone holding a cube with a computer showing it.

It is deliberately dumb. It hands out room codes, puts two sockets in a room,
and copies messages from one to the other. It never looks inside a `msg`
payload, keeps nothing on disk, and forgets a room as soon as both sides leave.

Protocol, all JSON text frames:

    in   {"t":"host"}                 ask for a room
         {"t":"join","code":"ABC234"} join one
         {"t":"msg","data":...}       send to the other side

    out  {"t":"code","code":"..."}    your room (host only)
         {"t":"joined","code":"..."}  you are in
         {"t":"peer","up":true|false} the other side arrived or left
         {"t":"msg","data":...}       from the other side
         {"t":"error","reason":"..."} and the socket closes
"""

import asyncio
import json
import os
import secrets
import time
from typing import Any, Optional

import websockets

# The name of the connection class moved between websockets releases, and the
# relay does not care which one it gets — anything with .send() and .close()
# will do. Keeping it untyped avoids pinning the server to one version.
Socket = Any

# No 0/O and no 1/I/L: these get read out loud and typed on a phone.
ALPHABET = "23456789ABCDEFGHJKMNPQRSTUVWXYZ"
CODE_LENGTH = 6

# A phone that locks its screen drops the socket. Hold the room open long
# enough for it to come back rather than making someone read the code again.
GRACE_SECONDS = 90
# A room nobody ever joins is a leak waiting to happen.
UNUSED_ROOM_SECONDS = 30 * 60
MAX_ROOMS = 500
# Blocks in one overlay, plus room to spare. An unbounded room is a way to
# spend somebody else's memory; nothing real comes close to this.
MAX_GUESTS = 8
MAX_MESSAGE_BYTES = 64 * 1024


class Room:
    """
    One host and everybody listening to it.

    It used to be one host and one guest, which is all the phone bridge needs
    and one fewer than the stream overlay needs. An overlay is a Browser Source
    per block — a clock, a solve list, a scramble — and with a single guest slot
    the second block to load silently took the first one's place, so the first
    sat there waiting for a trainer that was right in front of it.

    Messages go from anyone to everyone else, which leaves the bridge behaving
    exactly as before: with two sockets in the room, "everyone else" is one.
    """

    __slots__ = ("code", "host", "guests", "created", "empty_since", "max_guests")

    def __init__(self, code: str, max_guests: int = 1):
        self.code = code
        self.host: Optional[Socket] = None
        self.guests: set[Socket] = set()
        self.created = time.monotonic()
        self.empty_since: Optional[float] = None
        # How many may listen at once, which depends on what the room is FOR.
        # A bridge has exactly one phone: a second joining would put two cubes
        # into one app, which is why that was refused and still is. An overlay
        # is several blocks, each its own socket, and they are all read-only.
        self.max_guests = max_guests

    def others(self, ws: Socket) -> list[Socket]:
        out = [g for g in self.guests if g is not ws]
        if self.host is not None and self.host is not ws:
            out.append(self.host)
        return out

    def occupied(self) -> bool:
        return self.host is not None or bool(self.guests)


rooms: dict[str, Room] = {}


def new_code() -> str:
    while True:
        code = "".join(secrets.choice(ALPHABET) for _ in range(CODE_LENGTH))
        if code not in rooms:
            return code


async def send(ws: Socket, payload: dict) -> None:
    try:
        await ws.send(json.dumps(payload))
    except Exception:
        pass  # the socket is going away; the close handler will tidy up


async def fail(ws: Socket, reason: str) -> None:
    await send(ws, {"t": "error", "reason": reason})
    await ws.close()


async def handler(ws: Socket) -> None:
    room: Optional[Room] = None
    try:
        async for raw in ws:
            if isinstance(raw, bytes) or len(raw) > MAX_MESSAGE_BYTES:
                await fail(ws, "Message too large")
                return
            try:
                msg = json.loads(raw)
            except Exception:
                await fail(ws, "Bad message")
                return
            kind = msg.get("t")

            if kind == "host":
                if room is not None:
                    await fail(ws, "Already in a room")
                    return
                if len(rooms) >= MAX_ROOMS:
                    await fail(ws, "The relay is full, try again shortly")
                    return
                # The host says what the room is for. Absent means a bridge,
                # so a client that has never heard of this behaves as before.
                many = msg.get("kind") == "stream"
                room = Room(new_code(), MAX_GUESTS if many else 1)
                room.host = ws
                rooms[room.code] = room
                await send(ws, {"t": "code", "code": room.code})

            elif kind == "join":
                if room is not None:
                    await fail(ws, "Already in a room")
                    return
                code = str(msg.get("code", "")).upper()
                target = rooms.get(code)
                if target is None:
                    await fail(ws, "No computer is waiting on that code")
                    return
                if len(target.guests) >= target.max_guests:
                    # The wording matters for a bridge: a second phone trying to
                    # take a room is a mistake worth naming, not a full room
                    await fail(
                        ws,
                        "That room is full" if target.max_guests > 1 else "That code is already in use",
                    )
                    return
                room = target
                room.guests.add(ws)
                room.empty_since = None
                await send(ws, {"t": "joined", "code": room.code})
                if room.host is not None:
                    # The host hears about the first arrival only: for the
                    # bridge that is the one it cares about, and for an overlay
                    # "something is listening" is all it means either way
                    if len(room.guests) == 1:
                        await send(room.host, {"t": "peer", "up": True})
                    await send(ws, {"t": "peer", "up": True})

            elif kind == "msg":
                if room is None:
                    await fail(ws, "Not in a room")
                    return
                for other in room.others(ws):
                    await send(other, {"t": "msg", "data": msg.get("data")})

            else:
                await fail(ws, "Unknown message")
                return
    except Exception:
        pass
    finally:
        if room is not None:
            if room.host is ws:
                room.host = None
            else:
                room.guests.discard(ws)
            if room.occupied():
                # Only when the LAST listener goes: the host does not need to
                # hear "peer down" because one of five blocks was closed
                if room.host is not None and not room.guests:
                    await send(room.host, {"t": "peer", "up": False})
                for g in room.guests:
                    if room.host is None:
                        await send(g, {"t": "peer", "up": False})
            else:
                room.empty_since = time.monotonic()


async def sweep() -> None:
    """Forget rooms nobody is using."""
    while True:
        await asyncio.sleep(20)
        now = time.monotonic()
        for code, room in list(rooms.items()):
            if not room.occupied() and room.empty_since and now - room.empty_since > GRACE_SECONDS:
                rooms.pop(code, None)
            elif not room.guests and now - room.created > UNUSED_ROOM_SECONDS:
                if room.host is not None:
                    await fail(room.host, "Nobody joined; ask for a new code")
                rooms.pop(code, None)


async def main() -> None:
    host = os.environ.get("RELAY_HOST", "127.0.0.1")
    port = int(os.environ.get("RELAY_PORT", "8787"))
    async with websockets.serve(
        handler,
        host,
        port,
        ping_interval=20,
        ping_timeout=20,
        max_size=MAX_MESSAGE_BYTES,
    ):
        asyncio.create_task(sweep())
        print(f"cube relay listening on {host}:{port}", flush=True)
        await asyncio.Future()


if __name__ == "__main__":
    asyncio.run(main())

File: src/remote/relay.ts
/**
 * The socket to the relay, and the pairing dance on top of it.
 *
 * One side asks for a room and is given a code; the other side types that code
 * in. After that the relay is a pipe: whatever one side sends, the other
 * receives. It never inspects the contents.
 */

import type { BridgeMessage, FromRelay, ToRelay } from './protocol';

export type RelayStatus =
  /** Nothing going on */
  | 'idle'
  /** Opening the socket */
  | 'connecting'
  /** In a room, waiting for the other side */
  | 'waiting'
  /** Both sides present */
  | 'linked'
  /** Gave up; `error` says why */
  | 'error';

interface Listener {
  status?: (s: RelayStatus, code: string | null, error: string | null) => void;
  message?: (m: BridgeMessage) => void;
}

/** Where the relay lives. Same host as the app, so it inherits the certificate. */
export function relayUrl(): string {
  const override = import.meta.env?.VITE_RELAY_URL;
  if (override) return override;
  if (typeof location === 'undefined') return 'ws://localhost:8787';
  // The dev server does not proxy, so talk to the relay directly there
  if (location.port === '5173') return `ws://${location.hostname}:8787`;
  return `${location.protocol === 'https:' ? 'wss' : 'ws'}://${location.host}/relay`;
}

export class RelayLink {
  status: RelayStatus = 'idle';
  code: string | null = null;
  error: string | null = null;
  /** True once the other side has been seen, so a drop is worth retrying */
  private wasLinked = false;
  private ws: WebSocket | null = null;
  private listeners = new Set<Listener>();
  private retries = 0;
  private retryTimer: number | null = null;
  private role: 'host' | 'join' | null = null;
  private kind: 'bridge' | 'stream' = 'bridge';

  on(l: Listener): () => void {
    this.listeners.add(l);
    return () => this.listeners.delete(l);
  }

  private set(status: RelayStatus, error: string | null = null) {
    this.status = status;
    this.error = error;
    for (const l of [...this.listeners]) {
      try {
        l.status?.(status, this.code, error);
      } catch {
        /* a listener must not take the socket down */
      }
    }
  }

  /**
   * Ask for a room and a code to read out.
   *
   * `kind` says what the room is for, and the relay caps the listeners
   * accordingly: a bridge takes exactly one phone, because a second would put
   * two cubes into one app, while a stream overlay is several read-only blocks
   * and each is its own socket.
   */
  host(kind: 'bridge' | 'stream' = 'bridge') {
    this.role = 'host';
    this.kind = kind;
    this.code = null;
    this.open();
  }

  /** Join the room whose code was read out. */
  join(code: string) {
    this.role = 'join';
    this.code = code;
    this.open();
  }

  private open() {
    this.closeSocket();
    this.set('connecting');
    let ws: WebSocket;
    try {
      ws = new WebSocket(relayUrl());
    } catch (err) {
      this.set('error', (err as Error)?.message ?? 'Could not reach the relay');
      return;
    }
    this.ws = ws;

    ws.onopen = () => {
      this.retries = 0;
      this.send(this.role === 'host' ? { t: 'host', kind: this.kind } : { t: 'join', code: this.code! });
    };

    ws.onmessage = (ev) => {
      let msg: FromRelay;
      try {
        msg = JSON.parse(String(ev.data));
      } catch {
        return;
      }
      switch (msg.t) {
        case 'code':
          this.code = msg.code;
          this.set('waiting');
          break;
        case 'joined':
          this.code = msg.code;
          this.set('waiting');
          break;
        case 'peer':
          this.wasLinked = this.wasLinked || msg.up;
          this.set(msg.up ? 'linked' : 'waiting');
          break;
        case 'msg':
          for (const l of [...this.listeners]) {
            try {
              l.message?.(msg.data as BridgeMessage);
            } catch {
              /* one bad handler must not stop the stream */
            }
          }
          break;
        case 'error':
          this.set('error', msg.reason);
          this.closeSocket();
          break;
      }
    };

    ws.onclose = () => {
      if (this.status === 'error' || this.role === null) return;
      // A phone locking its screen kills the socket. The relay holds the room
      // open for a minute, so getting back in is usually just a reconnect.
      if (this.wasLinked && this.retries < 5) {
        this.retries++;
        this.set('connecting');
        this.retryTimer = window.setTimeout(() => this.open(), 500 * this.retries);
      } else {
        this.set('error', 'Connection to the relay was lost');
      }
    };

    ws.onerror = () => {
      // onclose always follows, and carries the retry logic
    };
  }

  private send(m: ToRelay) {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(JSON.stringify(m));
  }

  /** Pass something to the other side. */
  post(data: BridgeMessage) {
    this.send({ t: 'msg', data });
  }

  private closeSocket() {
    if (this.retryTimer) {
      clearTimeout(this.retryTimer);
      this.retryTimer = null;
    }
    const ws = this.ws;
    this.ws = null;
    if (!ws) return;
    ws.onopen = ws.onmessage = ws.onclose = ws.onerror = null;
    try {
      ws.close();
    } catch {
      /* already gone */
    }
  }

  stop() {
    this.role = null;
    this.wasLinked = false;
    this.retries = 0;
    this.closeSocket();
    this.code = null;
    this.set('idle');
  }
}

export const relay = new RelayLink();
